src/shared/ui/inputPrompter.ts

```diff
@@ -0,0 +1,109 @@
+    validateInput?(value: string): string | undefined
+}
+
+export type ExtendedInputBoxOptions = Omit<vscode.InputBoxOptions, 'buttons' | 'validateInput' | 'placeHolder'> &
+    AdditionalInputBoxOptions
+export type DataInputBox = Omit<vscode.InputBox, 'buttons'> & { buttons: PrompterButtons<string> }
+
+export const DEFAULT_INPUTBOX_OPTIONS: vscode.InputBoxOptions = {
+    ignoreFocusOut: true,
+}
+
+/**
+ * Creates a new InputBox.
+ *
+ * @param options Customizes the InputBox and InputBoxPrompter.
+ * @returns An InputBoxPrompter. This can be used directly with the `prompt` method or can be fed into a Wizard.
+ */
+export function createInputBox(options?: ExtendedInputBoxOptions): InputBoxPrompter {
+    const inputBox = vscode.window.createInputBox() as DataInputBox
+    applyPrimitives(inputBox, { ...DEFAULT_INPUTBOX_OPTIONS, ...options })
+
+    const prompter = new InputBoxPrompter(inputBox)
+
+    if (options?.validateInput !== undefined) {
+        prompter.setValidation(input => options.validateInput!(input))
+    }
+
+    return prompter
+}
+
+export class InputBoxPrompter extends Prompter<string> {
+    private _lastResponse?: string
+    private validateEvents: vscode.Disposable[] = []
+
+    public set lastResponse(response: string | undefined) {
+        this._lastResponse = typeof response === 'string' ? response : this._lastResponse
+        this.inputBox.value = this._lastResponse ?? ''
+    }
+
+    public get lastResponse(): string | undefined {
+        return this._lastResponse
+    }
+
+    constructor(public readonly inputBox: DataInputBox) {
+        super()
+    }
+
+    public setSteps(current: number, total: number): void {
+        this.inputBox.step = current
+        this.inputBox.totalSteps = total
+    }
+
+    public setValidation(validate: (value: string) => string | undefined): void {
+        this.validateEvents.forEach(d => d.dispose())
+        this.validateEvents = []
+
+        this.inputBox.onDidChangeValue(
+            value => (this.inputBox.validationMessage = validate(value)),
+            this.validateEvents
+        )
+        this.inputBox.onDidAccept(
+            () => (this.inputBox.validationMessage = validate(this.inputBox.value)),
+            this.validateEvents
+        )
+    }
+
+    protected async promptUser(): Promise<PromptResult<string>> {
+        const promptPromise = new Promise<PromptResult<string>>(resolve => {
+            this.inputBox.onDidAccept(() => {
+                this.lastResponse = this.inputBox.value
+                if (!this.inputBox.validationMessage) {
+                    resolve(this.inputBox.value)
+                }
+            })
+            this.inputBox.onDidHide(() => resolve(undefined))
+            this.inputBox.onDidTriggerButton(button => {
+                if (button === vscode.QuickInputButtons.Back) {
+                    resolve(WIZARD_BACK)
+                } else if ((button as QuickInputButton<string>).onClick !== undefined) {
```

Comment:
`QuickInputButton` in this case is an extended interface. Would it make sense to name it something else? I'm trying to avoid excess verbosity but also keep the names similar to the underlying APIs.

Reply:
What I'm trying to get at is why isn't this in terms of `WizardButton`? Seems like WizardButton isn't really being used...what value does it provide if you're not using it when you're actually dealing with the buttons themselves?

Reply:
So you're correct that `WizardButton` isn't being used outside the class itself. I'm using it as an intermediate type to describe higher-order types. But remember that these types are aliases for the underlying structure of an object; their names themselves have no meaning except for what we give them. The `PrompterButtons` type would function the exact same if we just substituted the type alias with its definition. But I think it reads cleaner by defining intermediate types to construct new types, just as one defines local variables to compose the final value of a function body.